File: 45. Jump Game II/main.py
import sys
from typing import List


class Solution:
    def jump(self, nums: List[int]) -> int:
        size = len(nums)
        jumps = [sys.maxsize] * size

        jumps[0] = 0
        for i in range(size):
            for j in range(i + 1, min(nums[i] + i + 1, size)):
                jumps[j] = min(jumps[j], jumps[i] + 1)

        return jumps[size - 1]


class Solution:
    def jump(self, nums: List[int]) -> int:
        size = len(nums)
        if size <= 1:
            return 0

        cur_farthest = 0
        next_farthest = 0
        jumps = 0

        for i in range(size):
            next_farthest = max(next_farthest, i + nums[i])
            if i == cur_farthest:
                jumps += 1
                cur_farthest = next_farthest
                if cur_farthest >= size - 1:
                    break

        return jumps


class Solution:
    def jump(self, nums: List[int]) -> int:
        size = len(nums)
        jumps = 0
        left = 0
        right = 0

        while right < size - 1:
            far_reach = 0
            for i in range(left, right + 1):
                far_reach = max(far_reach, i + nums[i])

            left = right + 1
            right = far_reach
            jumps += 1

        return jumps

File: 45. Jump Game II/main.rs
struct Solution;

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let size = nums.len();
        let mut jumps = vec![i32::MAX; size];

        jumps[0] = 0;
        for i in 0..size {
            for j in i + 1..size.min(nums[i] as usize + i + 1) {
                jumps[j] = jumps[j].min(jumps[i] + 1);
            }
        }

        jumps[size - 1]
    }
}

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let size = nums.len();
        let mut left = 0;
        let mut right = 0;
        let mut jumps = 0;

        while right < size - 1 {
            let mut far_reach = 0;
            for i in left..=right {
                far_reach = far_reach.max(nums[i] as usize + i);
            }

            left = right + 1;
            right = far_reach;
            jumps += 1;
        }

        return jumps;
    }
}

impl Solution {
    pub fn jump(nums: Vec<i32>) -> i32 {
        let size = nums.len();
        if size <= 1 {
            return 0;
        }

        let mut cur_farthest = 0;
        let mut next_farthest = 0;
        let mut jumps = 0;

        for i in 0..size {
            next_farthest = next_farthest.max(nums[i] as usize + i);
            if i == cur_farthest {
                jumps += 1;
                cur_farthest = next_farthest;
                if cur_farthest >= size - 1 {
                    break;
                }
            }
        }

        jumps
    }
}
